scheds/rust/scx_rustlite/src/bpf/main.bpf.c

```diff
@@ -0,0 +1,370 @@
+void BPF_STRUCT_OPS(rustlite_enqueue, struct task_struct *p, u64 enq_flags)
+{
+        struct task_ctx task;
+
+	/*
+	 * Scheduler is dispatched directly in .dispatch() when needed, so
+	 * we can skip it here.
+	 */
+	if (is_usersched_task(p))
+		return;
+
+        /*
+	 * Dispatch per-cpu kthreads on the local FIFO directly from the
+	 * kernel.
+         */
+	if (is_kthread(p) && p->nr_cpus_allowed == 1) {
+		bpf_printk("dispatch (local): pid=%d", p->pid);
+		dispatch_task_local(p, enq_flags | SCX_ENQ_LOCAL);
+		return;
+	}
+
+	/*
+	 * Other tasks can be added to the @enqueued list and they will be
+	 * processed by the user-space scheduler.
+	 *
+	 * If the @enqueued list is full (user-space scheduler is congested)
+	 * tasks will be dispatched directly from the kernel to the global
+	 * FIFO.
+	 */
+	task.pid = p->pid;
+	task.vtime = task_vtime(p);
+	bpf_printk("enqueue: pid=%d vtime=%llu", task.pid, task.vtime);
+	if (bpf_map_push_elem(&enqueued, &task, 0)) {
+		bpf_printk("dispatch (global): pid=%d", task.pid);
+		dispatch_task_global(p, enq_flags);
+		return;
+	}
+	__sync_fetch_and_add(&nr_enqueues, 1);
+
+	/*
+	 * Task was sent to user-space correctly, now we can wake-up the
+	 * user-space scheduler.
+	 */
+	usersched_needed = true;
+}
+
+/* Dispatch the user-space scheduler */
+static void dispatch_user_scheduler(void)
+{
+	struct task_struct *p;
+
+	if (!usersched_needed)
+		return;
+	usersched_needed = false;
+
+	p = bpf_task_from_pid(usersched_pid);
+	if (!p) {
+		scx_bpf_error("Failed to find usersched task %d", usersched_pid);
+		return;
+	}
+	bpf_printk("dispatch (scheduler): pid=%d", usersched_pid);
+	scx_bpf_dispatch(p, SCX_DSQ_GLOBAL, slice_ns, 0);
+	bpf_task_release(p);
+}
+
+/* Dispatch tasks that are ready to run */
+void BPF_STRUCT_OPS(rustlite_dispatch, s32 cpu, struct task_struct *prev)
+{
+	/* Check if the user-space scheduler needs to run */
+	dispatch_user_scheduler();
+
+	/*
+	 * Consume all tasks from the dispatched list and immediately dispatch
+	 * them to the global FIFO (the proper ordering has been already
+	 * determined by the user-space scheduler).
+	 *
+	 * A trace of the dispatched tasks can be seen in
+	 * /sys/kernel/debug/tracing/trace_pipe.
+	 */
+	bpf_repeat(MAX_ENQUEUED_TASKS) {
+		struct task_struct *p;
+		struct task_ctx task;
+
+		if (!scx_bpf_dispatch_nr_slots())
+			break;
+		if (bpf_map_pop_elem(&dispatched, &task))
+			break;
+		p = bpf_task_from_pid(task.pid);
+		if (!p)
+			continue;
+		bpf_printk("dispatch: pid=%d vtime=%llu", task.pid, task.vtime);
+		dispatch_task_global(p, 0);
+		bpf_task_release(p);
+	}
+}
+
+/* Select the target CPU for a task that is ready to run */
+s32 BPF_STRUCT_OPS(rustlite_select_cpu, struct task_struct *p, s32 prev_cpu,
+		   u64 wake_flags)
+{
+	s32 cpu;
+
+	/*
+	 * Keep using the same CPU if the task can only run there or if the CPU
+	 * is idle.
+	 */
+	if (p->nr_cpus_allowed == 1 ||
+	    scx_bpf_test_and_clear_cpu_idle(prev_cpu))
```

Comment:
So here (and below), we're grabbing an idle core, but then not dispatching the task to that idle core when it's enqueued in `rustlite_enqueue()` if it's not a per-cpu kthread. This is a bit sub-optimal because we don't really gain anything from nabbing that idle core, and we're taking it away from another task that could have come along and grabbed that core and enqueued directly on it.

If you look at [`scx_userland.bpf.c`](https://github.com/sched-ext/scx/blob/main/scheds/c/scx_userland.bpf.c#L113), we only do anything in the `ops.select_cpu()` callback if that task is going to be scheduled in the kernel directly, so we may want to do something similar here. Or, alternatively, it might be a reasonable policy to always schedule a task directly in the kernel if you're able to find an idle core.

Reply:
Ahh ok, this part was not very clear to me. Now, IIUC in `ops.select_cpu()` we can return a CPU that is where `ops.enqueue()` will be called for the task, right? That would explain the force_local and the dispatch from `ops.enqueue()`.

If that's right I'll mimic what scx_userland does and I'll try to clarify all of this with some comments.

Reply:
Yes, correct. The CPU returned by ops.select_cpu() is where the task will be migrated to and enqueued on in ops.enqueue(). So exactly as you're saying, the force local, etc is how we communicate that the task should enqueue locally on that DSQ.

@htejun, perhaps it would be useful to always set the SCX_ENQ_LOCAL flag any time a task is migrated to a different CPU from a call to ops.select_cpu()?

Reply:
Wouldn't that be a bit too magical? SCX_ENQ_LOCAL itself is a quirky thing which is meaningful only when the default select_cpu is used. It is tempting to expand it to make it more useful but the benefit is pretty trivial & marginal, I think. e.g. your suggested semantics can be really confusing if an implementation switches from default select_cpu implementation to custom one. If the enqueue path is not updated, it'll work as expected sometimes, which is a pretty bad failure mode. The flag name becomes a misnomer too.

I wonder what we should do is making it really clear that the flag is only meaningful when the default select_cpu is used.

Reply:
> SCX_ENQ_LOCAL itself is a quirky thing which is meaningful only when the default select_cpu is used.

The fact that it's only useful when default select_cpu is used is really odd to me and seems circumstantial, considering that it's a part of the "public" API that's consumable and used by schedulers. To me, what it really should mean is, "This is a hint (that you can ignore) that you might want to enqueue this on the local DSQ". In my opinion we should either document the semantics to make it consistent for any implementation of the scheduler, or we should get rid of it altogether.

> If the enqueue path is not updated, it'll work as expected sometimes, which is a pretty bad failure mode.

I'm not sure I understand this point. What do you mean by "not updated"?

Reply:
> The default select_cpu() needs a way to communicate it.

Right, but IMO the relevant point is that the flag does communicate _something_ about what CPU was chosen in ops.select_cpu(). The fact that it's only necessary to accommodate the default implementation of ops.select_cpu() is an implementation detail that we shouldn't be leaking. Put in other words, right now `SCX_ENQ_LOCAL` means: "you didn't implement ops.select_cpu(), and this is an (opaque) hint that you may want to enqueue locally". I think the flag should be made more generic than that. Something like `SCX_CPU_CHANGED`. The default implementation can and needs to use it, but there's no reason we shouldn't make it available to schedulers that implement ops.select_cpu() and want the same semantics for convenience.

>The fact that we're providing the default select_cpu implementation is where the quirkiness is coming from. The usefulness to quirkiness factor doesn't seem too bad to me.

Eh, I don't really agree. The existence of that flag adds surface area to the API, which itself adds cognitive load to users trying to understand how things work. IMO it's pretty non-intuitive that the flag is only relevant for if you don't implement ops.select_cpu(), and given how likely infrequent it is that someone won't implement ops.select_cpu(), it feels like it provides minimal value.

To be honest, IMO we shouldn't even be tryind to find idle cores in ops.select_cpu(). The default behavior should probably just be to return `prev_cpu`, as I'd expect the default implementation to do the minimal, sane thing. Selecting idle cores isn't a good idea for every use case, but `prev_cpu` is consistent behavior regardless of host topology.

>The meaning of the flag changes between default select_cpu and non-default ones, which can cause a lot of confusion. We can be strict about it and e.g. have a kfunc interface instead so that there's an explicit query for "did the default select_cpu indicated local enqueue?" which triggers ops_error if default select_cpu is not enabled, but that seems like an overkill to me.

I agree that it's overkill and a poor UX to add a kfunc for this. The `enq_flags` field feels like it's meant for this exact scenario. I guess I'm just not understanding why we can't change it to be generic and apply to both the default and non-default uses?

Reply:
I have a hard time understanding what you're suggesting because SCX_ENQ_CPU_CHANGED doesn't mean the same thing as what SCX_ENQ_LOCAL means. How would that work in practice?

As for the default implementation of select_cpu() becoming simpler, I'm not sure. It can be but I'm not sure whether or how much that'd be better. It's nice to have an effective default behavior as long as it's generic so that even really simple schedulers can be useful and we can further optimize the default behavior to e.g. make it NUMA area and whatnot.

As for cognitive overhead, yes it does add a bit but it's a small one. Maybe we can rename the flag to be clearer but it's something which is easily addressed by documentation and examples. Of the learning curves that writing a sched_ext scheduler involves, this isn't a noticeable one.

API cleanliness is a factor that we need to consider but not the only one and it's easy to overdo. Practically speaking, the impact here one way or the other is negligible, no?

Reply:
Sorry, let me try to clarify what I mean a bit.

>I have a hard time understanding what you're suggesting because SCX_ENQ_CPU_CHANGED doesn't mean the same thing as what SCX_ENQ_LOCAL means. How would that work in practice?

So my issue is that `SCX_ENQ_LOCAL` feels like a weird and leaky part of the API to me because it doesn't actually communicate any _concrete_ state / signal. It's a _hint_ that it's advisable to enqueue the task on the local dsq of @p's rq, but how am I supposed to really interpret that if I'm implementing my own ops.enqueue()? Wouldn't it be more useful to have flags like:

```
/*
 * The task has changed CPUs since the last time it was enqueued.
 */
SCX_ENQ_CPU_CHANGED = 1LLU << 42,
```

A BPF prog could then do `if ((enq_flags & SCX_ENQ_CPU_CHANGED) && (enq_flags & SCX_ENQ_WAKEUP))` to know that the task was migrated at wakeup time, and could dispatch it locally if it wants to. If a task is migrated to a different on say the dispatch path, then `(enq_flags & SCX_ENQ_WAKEUP)` wouldn't be set. A prog could choose to do something with that information, or it could choose to ignore it. The point is that it actually has concrete signal on the state of the task, rather than having to decide if it wants to listen to the opaque hint about whether to enqueue locally.

>As for cognitive overhead, yes it does add a bit but it's a small one. Maybe we can rename the flag to be clearer but it's something which is easily addressed by documentation and examples. Of the learning curves that writing a sched_ext scheduler involves, this isn't a noticeable one.

Yeah I agree it's not horrible cognitive load. If there ended up being a non-negligible complexity tradeoff to implement this, I agree there's an argument to be made against it.

>API cleanliness is a factor that we need to consider but not the only one and it's easy to overdo. Practically speaking, the impact here one way or the other is negligible, no?

I agree that API cleanliness isn't ironclad. I think my confusion is that I'm not quite understanding what the issue is with making the flag more generic. Is there some complexity that I'm overlooking?

Reply:
What's confusing to me is that `SCX_ENQ_LOCAL` doesn't indicate that the CPU has changed from `@prev_cpu`. It is telling `.enqueue()` that the CPU the task is on at the time of enqueue is ready for the task immediately and thus the enqueueing operation can be optimized into direct dispatch on the CPU. `SCX_ENQ_LOCAL` can be set when the task stayed on `@prev_cpu` too, so I don't see how "cpu changed" flag can be a substitute. This is `.force_local` flag for the default select_cpu implementation and means exactly the same thing and that doesn't have anything to do with whether the task's rq has changed or not.